### PR TITLE
Tests: Refactor iframe tests for async ready event

### PR DESCRIPTION
**Pull request by dmethvin**
Ref gh-3040

~~Not really happy with this solution for the long term, it seems like should be
having the iframe code signal when it's done. The old code would have broken for
any use of Deferred or setTimeout in the iframe, and it's likely this will break
too.~~ See below.

### CODE REVIEW

**Comment by mgol** [comment]
This seems pretty fragile...

**Comment by dmethvin** [comment]
Completely agree. The better solution is to have the iframe call back to let the parent know it's done. That requires some (simple I think) changes to all the tests run via `testIframe` but I didn't have the time to think through and test that.

We can either merge this to make the problem go away now or someone can do the right thing when they get a chance. (Or revert the async ready code I guess.)

**Comment by gibson042** [comment]
There are only four uses of `testIframe`, and one of them is in documentation:
* [large document dimensions](https://github.com/jquery/jquery/blob/055cb7534e2dcf7ee8ad145be83cb2d74b5331c7/test/unit/dimensions.js#L472)
* a dubious test of [multiple instance independence](https://github.com/jquery/jquery/blob/055cb7534e2dcf7ee8ad145be83cb2d74b5331c7/test/unit/selector.js#L492)
* [HTML5 boolean attributes](https://github.com/jquery/jquery/blob/055cb7534e2dcf7ee8ad145be83cb2d74b5331c7/test/unit/selector.js#L292)
* [README](https://github.com/jquery/jquery/blob/055cb7534e2dcf7ee8ad145be83cb2d74b5331c7/README.md#load-tests-in-an-iframe)

Especially considering that ready is an excludable module, I'd like to do things the right way and drop all four, replacing `testIframe` with `testIframeWithCallback` (which is almost identical anyway). Of those tests, only the first even plausibly cares about document readiness.

**Comment by dmethvin** [comment]
Okay, since @gibson042 did the searching and I have about 90 minutes now, imma try to get this fixed the right way. Hold my coat I'm going in!

**Comment by timmywil** [comment]
@dmethvin I assume someone got your coat. Are you taking anything else off?

**Comment by dmethvin** [comment]
WHY ARE THE FIRST TWO ARGS REVERSED FOR THESE GUYS???!

**Comment by dmethvin** [comment]
WHY DOES testIframe add the `.html` but testIframeWithCallback NOT????

**Comment by dmethvin** [comment]
What a mess! There were a bunch of other `testIframe()` tests in offset.js but I think I got everything.

**Comment by gibson042** [comment]
Huh, I don't know how I missed those. At any rate, thanks @dmethvin. You are a gentleman and a scholar.

File: README.md [hunk 1]
```diff
@@ -340,32 +340,25 @@ url("data/test.php?foo=bar");
...
-testIframe( fileName, testName, callback );
-```
-
-Callback arguments:
+### Load tests in an iframe (window.iframeCallback) ###
 
 ```js
-callback( jQueryFromIFrame, iFrameWindow, iFrameDocument );
+testIframeWithCallback( testName, fileName,
+  function callback( arg1, arg2, ... assert ) {
```
Comment (mgol) [inline README.md]: This syntax looks weird to me. Did you mean:
```js
function callback( arg1, arg2, ..., assert ) {
	...
} );
```
?

**Comment by mgol** [comment]
That's great! :clap:

**Comment by dmethvin** [comment]
I added another commit to clean up a bunch of things. Now each iframe page includes a file that defines `startIframeTest()` which is what the iframe calls to get the test started. The callback function gets `assert` as its first arg rather than it being tacked onto the end, that matches `QUnit.test()` conventions and may be a little less error prone. If this looks good to everyone I'd like to land it tomorrow before it has to be rebased.

File: test/data/iframeTest.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// Just used in iframed html tests which jshint doesn't see
+// jshint unused: false
```
Comment (markelog) [inline test/data/iframeTest.js]: `window.startIframeTest = function() {}` should help with jshint?

File: README.md [hunk 3]
```diff
@@ -340,32 +340,33 @@ url("data/test.php?foo=bar");
...
-Loads a given page constructing a url with fileName: `"./data/" + fileName + ".html"`
-and fires the given callback on jQuery ready (using the jQuery loading from that page)
-and passes the iFrame's jQuery to the callback.
+Some tests may require a document other than the standard test fixture, and
+these can be run in a separate iframe. The actual test code and assertions
+remain in jQuery's main test files; only the minimal test fixture markup
+and setup code should be placed in the iframe file.
 
 ```js
-testIframe( fileName, testName, callback );
```
Comment (markelog) [inline README.md]: Could this be done as `QUnit` plugin?

Comment (dmethvin) [inline README.md]: Possibly? I don't think it will be nicely encapsulated because the iframe test does not include QUnit.

**Pull request closed without merging**